src/commands/playground/util.rs

```diff
@@ -198,26 +197,44 @@ pub fn hoise_crate_attributes(code: &str, after_crate_attrs: &str, after_code: &
 /// Utility used by the commands to wrap the given code in a `fn main` if not already wrapped.
 /// To check, whether a wrap was done, check if the return type is Cow::Borrowed vs Cow::Owned
 /// If a wrap was done, also hoists crate attributes to the top so they keep working
-pub fn maybe_wrap(code: &str, result_handling: ResultHandling) -> Cow<'_, str> {
-	if code.contains("fn main") || code.contains("#![no_main]") {
-		return Cow::Borrowed(code);
-	}
+pub fn maybe_wrap(code: &str, result_handling: ResultHandling) -> String {
+	use quote::quote;
+	use syn::{parse::Parse, *};
 
-	// fn main boilerplate
-	let after_crate_attrs = match result_handling {
-		ResultHandling::None => "fn main() {\n",
-		ResultHandling::Discard => "fn main() { let _ = {\n",
-		ResultHandling::Print => "fn main() { println!(\"{:?}\", {\n",
-	};
+	struct Inline {
+		attrs: Vec<Attribute>,
+		exprs: Vec<Stmt>,
+	}
 
-	// fn main boilerplate counterpart
-	let after_code = match result_handling {
-		ResultHandling::None => "}",
-		ResultHandling::Discard => "}; }",
-		ResultHandling::Print => "}); }",
-	};
+	impl Parse for Inline {
+		fn parse(input: parse::ParseStream) -> Result<Self> {
+			let attrs = Attribute::parse_inner(input)?;
+			let exprs = Block::parse_within(input)?;
+			for stmt in &exprs {
+				if let Stmt::Item(Item::Fn(ItemFn { sig, block, .. })) = stmt {
+					if sig.ident.to_string() == "main" && sig.inputs.len() == 0 {
+						return Ok(Self {
+							attrs,
+							exprs: block.stmts.clone(),
+						});
+					}
+				}
+			}
+			Ok(Self { attrs, exprs })
+		}
+	}
 
-	Cow::Owned(hoise_crate_attributes(code, after_crate_attrs, after_code))
+	let Inline { attrs, exprs } = parse_str::<Inline>(code).unwrap();
+	let attrs = attrs.into_iter().map(|x| quote!(#x));
+	let exprs = exprs.into_iter().map(|x| quote!(#x));
+	match result_handling {
+		ResultHandling::None => quote! { #(#attrs)* fn main() { #(#exprs)* } },
+		ResultHandling::Discard => quote! { #(#attrs)* fn main() { _ =  { #(#exprs)* } } },
+		ResultHandling::Print => {
+			quote! { #(#attrs)* fn main() { println!("{:?}", { #(#exprs)* }) } }
+		}
+	}
+	.to_string()
```

Comment:
⚠️ Hmm, this could be quite problematic as well, now that we think of it. Indeed, it will mess up whitespace and the diagnostics won't map properly to what the user wrote on Discord.

We'll probably need to be using `format!`s and thread through the `str` inputs rather than just `quote!` 😬:

 1. parse the `&str` into `attrs` and `stmts`;
 2. binary search in the `&str` a valid splitting point so that `parse_str::<Attrs>(&code[..idx]).to_string()` (sort to speak) and our `attrs` do match.
 3. Define `let (attrs, stmts) = code.split_at(idx);`
 4. Replace `quote!` with `format!`s, using the `attrs`, `stmts` above

Reply:
the thing is though, the current code already doesnt handle fixing spans? and, if its just one line, your spans will have to be broken?

Reply:
I'll merge as is for now and if we need to fix it later, it's ok to make a new PR.

Reply:
:( this was a very significant usability regression, way worse than the original bug